File: src/main/java/org/moinex/ui/dialog/wallettransaction/AddTransferController.java
/*
 * Filename: AddTransferController.java
 * Created on: October  4, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.wallettransaction;

import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.error.MoinexException;
import org.moinex.model.Category;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.CalculatorService;
import org.moinex.service.CategoryService;
import org.moinex.service.I18nService;
import org.moinex.service.WalletService;
import org.moinex.service.WalletTransactionService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Add Transfer dialog
 */
@Controller
@NoArgsConstructor
public class AddTransferController extends BaseTransferManagement {
    @Autowired
    public AddTransferController(
            WalletService walletService,
            WalletTransactionService walletTransactionService,
            CalculatorService calculatorService,
            CategoryService categoryService,
            I18nService i18nService,
            ConfigurableApplicationContext springContext) {
        super(
                walletService,
                walletTransactionService,
                calculatorService,
                categoryService,
                i18nService,
                springContext);
    }

    public void setSenderWalletComboBox(Wallet wt) {
        if (wallets.stream().noneMatch(w -> w.getId().equals(wt.getId()))) {
            return;
        }

        senderWalletComboBox.setValue(wt);
        updateSenderWalletBalance();
    }

    public void setReceiverWalletComboBox(Wallet wt) {
        if (wallets.stream().noneMatch(w -> w.getId().equals(wt.getId()))) {
            return;
        }

        receiverWalletComboBox.setValue(wt);
        updateReceiverWalletBalance();
    }

    @Override
    @FXML
    protected void initialize() {
        super.initialize();
    }

    @Override
    @FXML
    protected void handleSave() {
        Wallet senderWt = senderWalletComboBox.getValue();
        Wallet receiverWt = receiverWalletComboBox.getValue();
        String transferValueString = transferValueField.getText();
        String description = descriptionField.getText();
        LocalDate transferDate = transferDatePicker.getValue();

        Category category = categoryComboBox.getValue();

        if (senderWt == null
                || receiverWt == null
                || transferValueString == null
                || transferValueString.isBlank()
                || description == null
                || description.isBlank()
                || transferDate == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_MESSAGE));
            return;
        }

        try {
            BigDecimal transferValue = new BigDecimal(transferValueString);

            LocalTime currentTime = LocalTime.now();
            LocalDateTime dateTimeWithCurrentHour = transferDate.atTime(currentTime);

            walletTransactionService.transferMoney(
                    senderWt.getId(),
                    receiverWt.getId(),
                    category,
                    dateTimeWithCurrentHour,
                    transferValue,
                    description);

            WindowUtils.showSuccessDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_TRANSFER_CREATED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_TRANSFER_CREATED_MESSAGE));

            Stage stage = (Stage) descriptionField.getScene().getWindow();
            stage.close();
        } catch (NumberFormatException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_VALUE_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_VALUE_MESSAGE));
        } catch (MoinexException.SameSourceDestinationException
                | IllegalArgumentException
                | EntityNotFoundException
                | IllegalStateException
                | MoinexException.InsufficientResourcesException
                | MoinexException.TransferFromMasterToVirtualWalletException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_ERROR_CREATING_TRANSFER_TITLE),
                    e.getMessage());
        }
    }

    /**
     * Updates the projected balance of a wallet (sender or receiver) after a transfer,
     * considering the rules for virtual and master wallets.
     *
     * @param currentWallet The wallet whose balance is being calculated (sender or receiver)
     * @param otherWallet   The other wallet involved in the transaction
     * @param label         The UI Label to be updated
     * @param isSender      True if the currentWallet is the sender, false if it is the receiver
     */
    @Override
    protected void updateAfterBalance(
            Wallet currentWallet, Wallet otherWallet, Label label, boolean isSender) {
        String transferValueString = transferValueField.getText();

        if (transferValueString == null
                || transferValueString.isBlank()
                || currentWallet == null
                || currentWallet.equals(otherWallet)) {
            UIUtils.resetLabel(label);
            return;
        }

        try {
            BigDecimal transferValue = new BigDecimal(transferValueString);

            if (transferValue.compareTo(BigDecimal.ZERO) < 0) {
                UIUtils.resetLabel(label);
                return;
            }

            BigDecimal afterBalance;

            if (isSender) {
                if (currentWallet.isMaster()
                        && otherWallet.isVirtual()
                        && otherWallet.getMasterWallet().equals(currentWallet)) {
                    // If the current wallet is the SENDER and is a master wallet of the receiver,
                    // then an error is thrown... BUT be a buddy and show a message to the user :)
                    WindowUtils.showInformationDialog(
                            i18nService.tr(
                                    Constants.TranslationKeys
                                            .WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_TITLE),
                            i18nService.tr(
                                    Constants.TranslationKeys
                                            .WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_MESSAGE));
                    return;
                }

                afterBalance = currentWallet.getBalance().subtract(transferValue);
            } else {
                // If the current wallet is the RECEIVER, the logic depends on its relationship with
                // the sender
                // Scenario: Virtual -> Linked Master Wallet
                // If the sender is a virtual wallet and the receiver is its master wallet,
                // the money is already in the total fund. It's just an internal reorganization
                if (otherWallet != null
                        && otherWallet.isVirtual()
                        && otherWallet.getMasterWallet().equals(currentWallet)) {
                    // The total balance of the master wallet does not change.
                    afterBalance = currentWallet.getBalance();
                } else {
                    // For all other cases (Master -> Master, Master -> Virtual, etc.),
                    // the receiver's balance simply increases.
                    afterBalance = currentWallet.getBalance().add(transferValue);
                }
            }

            if (afterBalance.compareTo(BigDecimal.ZERO) < 0) {
                UIUtils.setLabelStyle(label, Constants.NEGATIVE_BALANCE_STYLE);
            } else {
                UIUtils.setLabelStyle(label, Constants.NEUTRAL_BALANCE_STYLE);
            }

            label.setText(UIUtils.formatCurrency(afterBalance));

        } catch (NumberFormatException e) {
            UIUtils.resetLabel(label);
        }
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/wallettransaction/create/AddTransferController.kt
/*
 * Filename: AddTransferController.kt (original filename: AddTransferController.java)
 * Created on: October  4, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 21/03/2026
 */

package org.moinex.ui.dialog.wallettransaction.create

import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.stage.Stage
import org.moinex.common.constants.Constants
import org.moinex.common.constants.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.exception.MoinexException
import org.moinex.model.wallettransaction.Transfer
import org.moinex.model.wallettransaction.Wallet
import org.moinex.service.CalculatorService
import org.moinex.service.CategoryService
import org.moinex.service.PreferencesService
import org.moinex.service.wallet.WalletService
import org.moinex.ui.dialog.wallettransaction.base.BaseTransferManagement
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.time.LocalTime

@Controller
class AddTransferController(
    walletService: WalletService,
    calculatorService: CalculatorService,
    categoryService: CategoryService,
    preferencesService: PreferencesService,
    springContext: ConfigurableApplicationContext,
) : BaseTransferManagement(
        walletService,
        calculatorService,
        categoryService,
        preferencesService,
        springContext,
    ) {
    fun setSenderWalletComboBox(wt: Wallet) {
        if (wallets.none { it.id == wt.id }) {
            return
        }
        senderWalletComboBox.value = wt
        updateSenderWalletBalance()
    }

    fun setReceiverWalletComboBox(wt: Wallet) {
        if (wallets.none { it.id == wt.id }) {
            return
        }
        receiverWalletComboBox.value = wt
        updateReceiverWalletBalance()
    }

    @FXML
    override fun initialize() {
        super.initialize()
    }

    @FXML
    override fun handleSave() {
        val senderWt = senderWalletComboBox.value
        val receiverWt = receiverWalletComboBox.value
        val transferValueString = transferValueField.text
        val description = descriptionField.text
        val transferDate = transferDatePicker.value
        val category = categoryComboBox.value

        if (senderWt == null ||
            receiverWt == null ||
            transferValueString.isNullOrBlank() ||
            description.isNullOrBlank() ||
            transferDate == null
        ) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_MESSAGE),
            )
            return
        }

        runCatching {
            val transferValue = transferValueString.toBigDecimal()
            val currentTime = LocalTime.now()
            val dateTimeWithCurrentHour = transferDate.atTime(currentTime)

            walletService.createTransfer(
                Transfer(
                    senderWallet = senderWt,
                    receiverWallet = receiverWt,
                    date = dateTimeWithCurrentHour,
                    amount = transferValue,
                    description = description,
                    category = category,
                ),
            )

            WindowUtils.showSuccessDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_TRANSFER_CREATED_TITLE),
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_TRANSFER_CREATED_MESSAGE),
            )

            (descriptionField.scene.window as Stage).close()
        }.onFailure { e ->
            when (e) {
                is NumberFormatException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_VALUE_TITLE,
                        ),
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_VALUE_MESSAGE,
                        ),
                    )
                }
                is MoinexException.SameSourceDestinationException,
                is IllegalArgumentException,
                is EntityNotFoundException,
                is IllegalStateException,
                is MoinexException.InsufficientResourcesException,
                is MoinexException.TransferFromMasterToVirtualWalletException,
                -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_ERROR_CREATING_TRANSFER_TITLE,
                        ),
                        e.message ?: "Unknown error",
                    )
                }
                else -> throw e
            }
        }
    }

    override fun updateAfterBalance(
        mainWallet: Wallet?,
        otherWallet: Wallet?,
        afterBalanceLabel: Label,
        isSender: Boolean,
    ) {
        val transferValueString = transferValueField.text

        if (transferValueString.isNullOrBlank() ||
            mainWallet == null ||
            mainWallet == otherWallet
        ) {
            UIUtils.resetLabel(afterBalanceLabel)
            return
        }

        runCatching {
            val transferValue = transferValueString.toBigDecimal()

            if (transferValue < BigDecimal.ZERO) {
                UIUtils.resetLabel(afterBalanceLabel)
                return
            }

            val afterBalance = determineAfterBalance(mainWallet, otherWallet, transferValue, isSender)

            val style =
                if (afterBalance < BigDecimal.ZERO) {
                    Constants.NEGATIVE_BALANCE_STYLE
                } else {
                    Constants.NEUTRAL_BALANCE_STYLE
                }

            UIUtils.setLabelStyle(afterBalanceLabel, style)
            afterBalanceLabel.text = UIUtils.formatCurrency(afterBalance)
        }.onFailure {
            UIUtils.resetLabel(afterBalanceLabel)
        }
    }

    private fun determineAfterBalance(
        mainWallet: Wallet,
        otherWallet: Wallet?,
        transferValue: BigDecimal,
        isSender: Boolean,
    ): BigDecimal {
        return if (isSender) {
            if (mainWallet.isMaster() &&
                otherWallet?.isVirtual() == true &&
                otherWallet.masterWallet == mainWallet
            ) {
                WindowUtils.showInformationDialog(
                    preferencesService.translate(
                        TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_TITLE,
                    ),
                    preferencesService.translate(
                        TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_MESSAGE,
                    ),
                )
                return BigDecimal.ZERO
            }
            mainWallet.balance - transferValue
        } else {
            if (otherWallet?.isVirtual() == true &&
                otherWallet.masterWallet == mainWallet
            ) {
                mainWallet.balance
            } else {
                mainWallet.balance + transferValue
            }
        }
    }
}
